src/centralstore/incremental/NodeManager.cpp

```diff
@@ -0,0 +1,254 @@
+        node_manager_logger.error("Error while reading edge reference data from block " + std::to_string(blockAddress));
+    }
+
+    if (!NodeBlock::nodesDB->read(reinterpret_cast<char *>(&propRef), sizeof(unsigned int))) {
+        node_manager_logger.error("Error while reading prop reference data from block " + std::to_string(blockAddress));
+    }
+
+    if (!NodeBlock::nodesDB->read(&label[0], NodeBlock::LABEL_SIZE)) {
+        node_manager_logger.error("Error while reading label data from block " + std::to_string(blockAddress));
+    }
+    bool usage = usageBlock == '\1';
+    node_manager_logger.debug("Label = " + std::string(label));
+    node_manager_logger.debug("Length of label = " + std::to_string(strlen(label)));
+    node_manager_logger.debug("DEBUG: raw edgeRef from DB (disk) " + std::to_string(edgeRef));
+
+    nodeBlockPointer = new NodeBlock(nodeId, blockAddress, propRef, edgeRef, label, usage);
+
+    node_manager_logger.debug("DEBUG: nodeBlockPointer after creating the object edgeRef" +
+                              std::to_string(nodeBlockPointer->edgeRef));
+
+    if (nodeBlockPointer->edgeRef % RelationBlock::BLOCK_SIZE != 0) {
+        throw "Exception: Invalid edge reference address = " + nodeBlockPointer->edgeRef;
+    }
+    return nodeBlockPointer;
+}
+
+void NodeManager::persistNodeIndex() {
+    std::ofstream index_db(this->index_db_loc, std::ios::trunc | std::ios::binary);
+    if (index_db.is_open()) {
+        for (auto nodeMap : this->nodeIndex) {
+            char nodeIDC[NodeManager::INDEX_KEY_SIZE] = {0};  // Initialize with null chars
+            std::strcpy(nodeIDC, nodeMap.first.c_str());
+            index_db.write(nodeIDC, sizeof(nodeIDC));
+            unsigned int nodeBlockIndex = nodeMap.second;
+            index_db.write(reinterpret_cast<char *>(&(nodeBlockIndex)), sizeof(unsigned int));
+            node_manager_logger.debug("Writing node index --> Node key = " + std::string(nodeIDC) + " value " +
+                                      std::to_string(nodeBlockIndex));
+        }
+    }
+    index_db.close();
+}
+
+/**
+ * Return the number of nodes upto the limit given in the arg from nodes index
+ * Default limit is 10
+ * */
+std::list<NodeBlock> NodeManager::getGraph(int limit) {
+    int i = 0;
+    std::list<NodeBlock> vertices;
+    for (auto it : this->nodeIndex) {
+        i++;
+        if (i >= limit) {
+            break;
+        }
+        auto nodeId = it.first;
+        NodeBlock *node = this->get(nodeId);
+        vertices.push_back(*node);
+    }
+    return vertices;
+}
+
+void NodeManager::close() {
+    this->persistNodeIndex();
+    if (PropertyLink::propertiesDB) {
+        PropertyLink::propertiesDB->flush();
+        PropertyLink::propertiesDB->close();
+    }
+    if (NodeBlock::nodesDB) {
+        NodeBlock::nodesDB->flush();
+        NodeBlock::nodesDB->close();
+    }
+}
+
+const unsigned long NodeManager::INDEX_KEY_SIZE = 8;
```

Comment:
Can we have these constants defined in a header file?

Reply:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

Reply:
Done